File: 1973-count-nodes-equal-to-sum-of-descendants/1973-count-nodes-equal-to-sum-of-descendants.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int totalN=0;
    public int equalToDescendants_01(TreeNode root) {
        if(root==null) return 0;
        
        int leftSum = equalToDescendants_01(root.left);
        int rightSum = equalToDescendants_01(root.right);
        
        if(leftSum+rightSum == root.val) totalN++;
        
        return leftSum+rightSum+root.val;
    }
    public int equalToDescendants(TreeNode root) {
        equalToDescendants_01(root);
        return totalN;
    }
}

File: 1973-count-nodes-equal-to-sum-of-descendants/1973-count-nodes-equal-to-sum-of-descendants.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int totalN = 0;
    int count(TreeNode* node) {
        if(!node) return 0;
        
        long leftSum = count(node->left);
        long rightSum = count(node->right);
        
        if(leftSum + rightSum == node->val) totalN++;
        
        return rightSum+leftSum+node->val;
    }
    int equalToDescendants(TreeNode* root) {
        count(root);
        return totalN;
    }
};